plugin/pkg/shoot/quotavalidator/admission.go

```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
 	if a.GetKind().GroupKind() != garden.Kind("Shoot") {
 		return nil
 	}
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
 	shoot, ok := a.GetObject().(*garden.Shoot)
 	if !ok {
 		return apierrors.NewBadRequest("could not convert resource into Shoot object")
 	}
 
-	// TODO: implement quota limit check logic here
-	if false {
-		return admission.NewForbidden(a, fmt.Errorf("quota limits exceeded for %v", shoot.Name))
+	// Pass if the shoot is intended to get deleted
+	if shoot.DeletionTimestamp != nil && shoot.Annotations[common.ConfirmationDeletionTimestamp] != "" {
+		return nil
+	}
+
+	quotaReferences, err := h.getShootsQuotaReferences(*shoot)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	// Quotas are cumulative, means each quota must be not exceeded that the admission pass.
+	var maxShootLifetime *int
+	for _, quotaRef := range quotaReferences {
+		quota, err := h.quotaLister.
+			Quotas(quotaRef.Namespace).
+			Get(quotaRef.Name)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		// Get the max clusterLifeTime
+		if quota.Spec.ClusterLifetimeDays != nil {
+			if maxShootLifetime == nil {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+			if *maxShootLifetime > *quota.Spec.ClusterLifetimeDays {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+		}
+
+		exceededMetrics, err := h.isQuotaExceeded(*shoot, *quota)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+		if exceededMetrics != nil {
+			var message bytes.Buffer
```

Comment:
Any reason for a `Buffer` here?

Reply:
Buffer concatenation is more efficient (only O(n)) as concatenate strings directly. Think here we have at most six strings. Therefore switched to string concatenation.